### ISSUE TITLE
RealmChangeListener on RealmResults called even when data inside given collection is not modified.

### ISSUE THREAD

**Issue by mkoslacz** [realm/realm-java#2946]
#### Goal

I want to receive events only when a given `RealmObject` or a `RealmResults Collection` is altered.
#### Expected Results

A `RealmChangeListener` docs says:

>  Register against a `RealmResults` or `RealmObject` **to only get notified about changes to them**.
#### Actual Results

I got a notification on an every Realm change, just like in case of registering a `RealmChangeListener` against a `Realm`.
#### Steps & Code to Reproduce

Just add a `RealmChangeListener` to `RealmResults`, put logging there, and perform some changes not-connected with mentioned collection.
#### Code Sample

``` java

        RealmResults<Program> favourites = realm
                .where(Program.class)
                .greaterThan(Program.C_STARTTIME_INT, Timestamp.getNow())
                .lessThan(Program.C_STARTTIME_INT, Timestamp.getBegginingOfCurrentDay() +
                        Timestamp.getDayInSeconds())
                .equalTo(Program.C_USERRATING_INT, Program.USER_RATING_FAVOURITE)
                .findAllSorted(Program.C_STARTTIME_INT);

        favourites.addChangeListener(new RealmChangeListener<RealmResults<Program>>() {
            @Override
            public void onChange(RealmResults<Program> element) {
                Log.d(TAG, "favourites onChange: altered/new/removed elements size: " + element.size());
            }
        });
```

Results in following prints:

```
06-07 08:49:08.155 3701-3701/pl.wp.programtv D/FragmentCarousel: favourites onChange: altered/new/removed  elements size: 0
```
#### Version of Realm and tooling

Realm version(s): 1.0.0

Android Studio version: 2.1.1

Which Android version and device: Any, especially x86_64 emulator with Android 6.0 and Samsung SM-G531F with Android 5.1.1

**Comment by cmelchior** [comment]
Hi @mkoslacz 
How did you create the `favorites` ? Are you using the `findAllAsync` API's?

**Comment by mkoslacz** [comment]
Hi, @cmelchior 

Well, I've just updated the code sample. As you can see, I'm using a sync `findAllSorted`.

**Comment by cmelchior** [comment]
Then it is working as intended. The RealmChangeListener will also be triggered when the async result comes back. In your case, it came back empty.

**Comment by mkoslacz** [comment]
Actually, we update our data using a background thread. The question is, will every background Realm DB change call all RealmChangeListeners on our app?

**Comment by cmelchior** [comment]
Background commits will trigger the relevant listeners yes. If you commit a Foo object, listeners on Bar will not be notified, but any global RealmListeners will.

Right now we only operate with two types of listeners: Global and Type-based, so if you change any Foo object all listeners on Foo will be triggered. This is a bit more coarse grained than we would like, but it will be fixed as part of #989

**Comment by mkoslacz** [comment]
Ok, I see, thanks a lot. However, the documentation is a lil bit misleading right now.

**Comment by cmelchior** [comment]
Yes, I can see that the async case isn't described there. We should update that.

**Comment by cah-ericmaxwell** [comment]
Not sure if this is related, but we're seeing this behavior even when using a synchronous query.

We have something like the following...

`        this.myCustomSetting = realm.where(CustomSetting.class).equalTo("key", "DEVICE_IS_LOCKED").findFirst();
        this.myCustomSetting.addChangeListener(customSettingChangeListener);
`

`    RealmChangeListener customSettingChangeListener = new RealmChangeListener() {
        @Override
        public void onChange() {
             updateUi();
        }
    };
`

And have found that both of these 2 statements trigger the change listener.
`           r.beginTransaction();
            CustomSetting setting = new CustomSetting();
            setting.setKey("DEVICE_IS_LOCKED");
            setting.setValue("true");
            r.copyToRealmOrUpdate(setting);
            r.commitTransaction();
`

`           r.beginTransaction();
            CustomSetting setting = new CustomSetting();
            setting.setKey("DEVICE_LOCK_DATE");
            setting.setValue(new Date().toString());  
            r.copyToRealmOrUpdate(setting);
            r.commitTransaction();
`
But, I would not have expected the second update to fire a realm change on that object/listener because the Query filtered it by key equal to "DEVICE_IS_LOCKED".   Or is it that any update to the CustomSetting table would fire a change event?

**Comment by mkoslacz** [comment]
Hi again @cmelchior, I'm back with the issue. Now I'm using `Realm 1.2.0` and `2.0.0-SNAPSHOT`. If I edit mentioned sample to look like this:

```java

        RealmResults<Program> favourites = realm
                .where(Program.class)
                .greaterThan(Program.C_STARTTIME_INT, Timestamp.getNow())
                .lessThan(Program.C_STARTTIME_INT, Timestamp.getBegginingOfCurrentDay() +
                        Timestamp.getDayInSeconds())
                .equalTo(Program.C_USERRATING_INT, Program.USER_RATING_FAVOURITE)
                .findAllSorted(Program.C_STARTTIME_INT);

        favourites.addChangeListener(new RealmChangeListener<RealmResults<Program>>() {
            @Override
            public void onChange(RealmResults<Program> element) {
                Log.d(TAG, String.format("favourites onChange: altered/new/removed elements size: %d, whole list size: %d", element.size(), favourites.size()));
            }
        });
```

It reveals that the listener is called everytime when a background commit triggers, even when the given commit does not alter favourites. Each time log claims that whole list was changed:

```
06-07 08:49:08.155 3701-3701/pl.wp.programtv D/FragmentCarousel: favourites onChange: altered/new/removed  elements size: 162, whole list size: 162
```

Now i'm not able to stop my list from refreshing so the UI is flooded by refreshes.

**Comment by beeender** [comment]
@mkoslacz I cannot reproduce your problem. Is it possible you can create a sample project to reproduce it?

**Comment by mkoslacz** [comment]
@beeender @cmelchior yep, I have managed to create a sample project:
https://github.com/mkoslacz/RealmDataRefreshShowcase
I hope that it's readable enough, it's rather a quick draft.

On the `master` branch, using Realm 1.2.0, you can see two `todos` in which I described the problems:
-  saving x programs makes Realm save only one of them, what is visible on the `MainActivity.mAllPrograms` changeListener,
- saving programs makes Realm invoke the `MainActivity.mAllStations` changeListener which claims that all of them were modified.

On the `refreshing-fix` branch you can see that refreshing problem could be solved by removing some other models and linking to them. The problem is that actually I need this models in my project.

On the `saving-fix` branch you can see that redundant refreshes problem could be solved by bumping Realm version to `2.0.0-SNAPSHOT`. The problem is that it causes some development issues related with SNAPSHOT version api changes.

**Comment by mkoslacz** [comment]
@beeender, any updates on this? Can I provide you some improvements or explanations for my sample project?

**Comment by Zhuinden** [comment]
Every RealmChangeListener is called for a GIVEN TABLE when that table is modified. 

Change listener based on whether the results themselves have been modified are "fine-grained notifications" and they are tracked here   https://github.com/realm/realm-java/issues/989

It'll happen after the Results integration from the ObjectStore

**Comment by mkoslacz** [comment]
Yep, I know, but why is the Stations table change listener called on the `master` branch when the Programs table is updated and why does this behaviour stop on the `refreshing-fix` branch after removing some other tables? 

It looks like the Programs table and the Stations table aren't connected with each other in this use case. 

All of the model binding is performed in the `bind()` methods in each of this tables elements.

**Comment by beeender** [comment]
@mkoslacz I am terribly sorry about forgetting to update this issue ... Due to the way we implement the notifications on the collections, they would be false positive notifications triggered when non-related data changes. But it will be solved by #3834 and future PR for #989 . Those are coming soon. 
Very very sorry again for the late response.

**Comment by mkoslacz** [comment]
@beeender no problem, thanks for providing me these informations. Glad that improvements are coming, decreasing the number of refreshes will significantly improve our app performance. 
I'll stay tuned.

**Comment by beeender** [comment]
#3834 is merged to master. the false positive notification on `RealmResults` will be fixed in `v2.4.0`. But it will still have false positive notifications on `RealmObject` which is tracking by #4101 . Close this one.

**Comment by profause** [comment]
> @beeender no problem, thanks for providing me these informations. Glad that improvements are coming, decreasing the number of refreshes will significantly improve our app performance. I'll stay tuned.

@mkoslacz have you had any results on this yet? I have been trying to solve the same problem for some time now.
I assume you only want to be notified when the underlying data in a realmResult<?> changes not the entire result set.
